core/server/master/src/main/java/alluxio/master/meta/PathProperties.java

```diff
@@ -12,41 +12,235 @@
+    try (LockResource r = new LockResource(mLock.writeLock())) {
+      if (mState.getProperties().containsKey(path)) {
+        mState.applyAndJournal(ctx, RemovePathPropertiesEntry.newBuilder()
+            .setPath(path).build());
+      }
+    }
+  }
+
+  @Override
+  public Journaled getDelegate() {
+    return mState;
+  }
+
+  /**
+   * Journaled state of path level properties.
    */
-  public void setProperty(String path, PropertyKey key, String value) {
-    mProperties.computeIfAbsent(path, k -> new ConcurrentHashMap<>()).put(key, value);
+  @NotThreadSafe
+  public static final class State implements Journaled {
+    /**
+     * Map from path to key value properties.
+     */
+    private final Map<String, Map<String, String>> mProperties = new HashMap<>();
+
+    /**
+     * The map from paths to properties is unmodifiable, but for each path, the property map is
+     * modifiable.
+     *
+     * @return the internal properties
+     */
+    public Map<String, Map<String, String>> getProperties() {
+      return Collections.unmodifiableMap(mProperties);
+    }
+
+    @Override
+    public boolean processJournalEntry(Journal.JournalEntry entry) {
+      if (entry.hasPathProperties()) {
+        applyPathProperties(entry.getPathProperties());
+      } else if (entry.hasRemovePathProperties()) {
+        applyRemovePathProperties(entry.getRemovePathProperties());
+      } else {
+        return false;
+      }
+      return true;
+    }
+
+    private void applyPathProperties(PathPropertiesEntry entry) {
+      final String path = entry.getPath();
+      final Map<String, String> properties = entry.getPropertiesMap();
+      if (mProperties.containsKey(path)) {
+        mProperties.get(path).clear();
+        mProperties.get(path).putAll(properties);
+      } else {
+        mProperties.put(path, new HashMap<>(properties));
+      }
+    }
+
+    private void applyRemovePathProperties(RemovePathPropertiesEntry entry) {
+      final String path = entry.getPath();
+      mProperties.remove(path);
+    }
+
+    /**
+     * @param ctx the journal context
+     * @param entry the path properties entry
+     */
+    public void applyAndJournal(Supplier<JournalContext> ctx, PathPropertiesEntry entry) {
+      applyAndJournal(ctx, JournalEntry.newBuilder().setPathProperties(entry).build());
+    }
+
+    /**
+     * @param ctx the journal context
+     * @param entry the remove path properties entry
+     */
+    public void applyAndJournal(Supplier<JournalContext> ctx, RemovePathPropertiesEntry entry) {
+      applyAndJournal(ctx, JournalEntry.newBuilder().setRemovePathProperties(entry).build());
+    }
+
+    @Override
+    public void resetState() {
+      mProperties.clear();
+    }
+
+    @Override
+    public CheckpointName getCheckpointName() {
+      return CheckpointName.PATH_PROPERTIES;
+    }
+
+    @Override
+    public Iterator<Journal.JournalEntry> getJournalEntryIterator() {
```

Comment:
I think this could be significantly simplified with `Iterators.transform(mProperties.entrySet().iterator(), entry -> ...)`